Add `number` platform for LED brightness to air-Q

Sibgatulin opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Each air-Q device is equipped with 2 LED strips that function as LED level indicators / bar graphs for two (configurable) sensors. This PR exposes a single `number` entity for each `airq` device, allowing to set the brightness of both strips simultaneously.
[This](https://github.com/CorantGmbH/aioairq/compare/v0.4.4...v0.4.6) is the changes to the backend `aioairq` package that permit such brightness control.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [x] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/39129
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

Sibgatulin committed:
feat(airq): add `number` platform for LED brightness

Brightness is fetched by `AirQCoordinator` in `_async_update_data`
and is expressed in %

joostlek commented on homeassistant/components/airq/number.py:
```diff
@@ -0,0 +1,80 @@
+AIRQ_LED_BRIGHTNESS = NumberEntityDescription(
+    key="airq_led_brightness",
+    translation_key="airq_led_brightness",
+    native_min_value=0.0,
+    native_max_value=10.0,
+    native_step=0.1,
```
why don't we do 0-100 with a one step? That sounds more logical and then we can use PERCENTAGE as unit

Sibgatulin replied:
Fair point. The last commit I just pushed handles the conversion between the percentage and the [0, 10] scale expected by the firmware/`aioairq`. 
I will consider now if it makes sense to push this into `aioairq` and further bump the dependency...

Sibgatulin replied:
Thanks for looking into it all!

Sibgatulin replied:
@joostlek I did push these changes to `aioairq`, it now handles the brightness in percent by default. I reset and the aforementioned commit and force-pushed a new one.

joostlek commented on homeassistant/components/airq/number.py:
```diff
@@ -0,0 +1,80 @@
+    async_add_entities(entities)
+
+
+class AirQLEDBrightness(CoordinatorEntity, NumberEntity):
```
```suggestion
class AirQLEDBrightness(CoordinatorEntity[AirQCoordinator], NumberEntity):
```

Sibgatulin replied:
thanks, I missed this neat possibility

joostlek commented on homeassistant/components/airq/number.py:
```diff
@@ -0,0 +1,80 @@
+    ) -> None:
+        """Initialize a single sensor."""
+        super().__init__(coordinator)
+        self.entity_description: NumberEntityDescription = description
```
```suggestion
        self.entity_description = description
```

joostlek commented on homeassistant/components/airq/number.py:
```diff
@@ -0,0 +1,80 @@
+        self.async_write_ha_state()
+        await self._device.set_current_brightness(value)
+
+    @property
+    def _device(self) -> AirQ:
+        """Return the reference to the device API held by the coordinator."""
+        # the following assertion pacifies mypy:
+        assert isinstance(self.coordinator, AirQCoordinator)
+        return self.coordinator.airq
```
you only use this once, I am assuming you added this for mypy, but by adding the coordinator to the coordinatorEntity typeslot, this is now fixed

Sibgatulin replied:
you are completely right, thanks

joostlek commented on homeassistant/components/airq/strings.json:
```diff
@@ -35,6 +35,11 @@
   "entity": {
+    "number": {
+      "airq_led_brightness": {
+        "name": "LED Brightness"
```
```suggestion
        "name": "LED brightness"
```

Sibgatulin commented on homeassistant/components/airq/number.py:
```diff
@@ -0,0 +1,73 @@
+    @property
+    def native_value(self) -> float:
+        """Return the brightness of the LEDs in %."""
+        return float(self.coordinator.data.get("brightness", BRIGHTNESS_DEFAULT))
```
Good point, it is guaranteed to be there, this precaution is unnecessary

Sibgatulin commented on homeassistant/components/airq/number.py:
```diff
@@ -0,0 +1,73 @@
+        self._attr_device_info = coordinator.device_info
+        self._attr_unique_id = f"{coordinator.device_id}_{description.key}"
+
+    @property
+    def native_value(self) -> float:
+        """Return the brightness of the LEDs in %."""
+        return float(self.coordinator.data.get("brightness", BRIGHTNESS_DEFAULT))
+
+    async def async_set_native_value(self, value: float) -> None:
+        """Set the brightness of the LEDs to the value in %."""
+        _LOGGER.debug(
+            "Changing LED brighntess from %.1f%% to %.1f%%",
+            self.coordinator.data["brightness"],
+            value,
+        )
+        self.coordinator.data["brightness"] = value
+        self.async_write_ha_state()
+        await self.coordinator.airq.set_current_brightness(value)
```
Ok, thanks for pointing it out. I thought it was not strictly needed here where only one single description is implemented and intended. But I will then return to the pattern of lambdas in the description

Sibgatulin commented:
I believe, I addressed every comment, @joostlek. 
I also believe it would be better if I squashed the commits in this PR (since after this PR is rebased on top of https://github.com/home-assistant/core/pull/146169, some of the commits here reflect a broken/incorrect state of the code + many small commits with fixes). 
Are there any objections?

joostlek reviewed (changes requested):
As discussed on discord, the tests shouldn't touch hass.data and instead test the state via the state machine

Sibgatulin committed:
Factor common to airq tests data out into dedicated submodule

Sibgatulin committed:
Remove mock_setup_entry from airq.test_coordinator

I had added this `usefixtures("mock_setup_entry")` by accident

Sibgatulin committed:
Patch AirQ class in a pytest.fixture

Instead of explicitly patching each method, create a fixture that
patches the whole class with an AsyncMock and some meaningful defaults
for most used methods. Then patch only where and when needed

Sibgatulin committed:
Patch AirQ class in a pytest.fixture

Instead of explicitly patching calls to `aioairq.AirQ` methods,
patch out the whole class, as imported either in the coordinator or the
config flow, some of its most used methods with sensible defaults, and
then only modify the mocks when necessary

Sibgatulin committed:
Do not force float conversion in number entity

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>
